Skuba addon certificate auto rotation (#883)
SUSE/doc-caasp: adoc/admin-security-certificates.adoc

After this process, bootstrap the cluster with `skuba node bootstrap`.

=== Replace Server Certificate signed by a Trusted CA Certificate

{productname} uses `oidc-dex` and `oidc-gangway` servers to do authentication and authorization.
Administrators might choose to replace these server's certificates by issuing a trusted CA certificate after cluster deployment.
This way, the user does not have to add specific certificates to their trusted keychain.

* With The Trusted CA Key
+

> Please ref to <<addon_certificates_rotation>> on how to use cert-manager and reloader to issue `oidc-dex` and `oidc-gangway` certificate signed by trusted CA certificate/key.

Reviewer: ```suggestion
Please refer to <<addon_certificates_rotation>> on how to use cert-manager and reloader to issue `oidc-dex` and `oidc-gangway` certificate signed by trusted CA certificate/key.
```

* Without The Trusted CA Key
+
[WARNING]
====

> Due to the custom trusted CA certificate key is not in {kube} cluster.

Reviewer: ```suggestion
Because the custom trusted CA certificate is not in {kube} cluster, administrators must handle server certificate rotation manually before the certificate expires.
```

> Administrators must handle server certificate rotation manually before the certificate expires.

Reviewer: ```suggestion
```

Restarting the kubelet effectively rotates the certificate to read the latest
generated file.

[[addon_certificates_rotation]]
=== Addon certificates rotation

> The addon certificate could be automatically rotated by leverages the opensource solution `cert-manager` and `reloader`. The cert-manager is to auto rotates the addon certificates stored in Secret, and the reloader is to watch the addon certificate Secret and rolling upgrade the addon Deployments or DaemonSet.

Reviewer: ```suggestion
The addon certificates can be automatically rotated by leveraging the open-source solutions `cert-manager` and `reloader`. `cert-manager` is for automatically rotating the certificates stored in Secrets, and `reloader` is for watching and reconciling the updated Secrets to execute rolling upgrade the affecting addon Deployments or DaemonSet.
```

* Prerequisites

> . To let reloader auto rolling upgrade the addon Deployments or DaemonSet, we need to label the addons:

Reviewer: ```suggestion
. To let reloader automatically rolling upgrade the addon Deployments or DaemonSet, we need to label the addons:
```

kubectl annotate --overwrite deployment/oidc-gangway -n kube-system secret.reloader.stakater.com/reload=oidc-gangway-cert

kubectl annotate --overwrite deployment/metrics-server -n kube-system secret.reloader.stakater.com/reload=metrics-server-cert
----

. Upload the Kubernetes CA cert/key pair to Secret in namespace kube-system:
+
[source,bash]
----
kubectl create secret tls kubernetes-ca --cert=pki/ca.crt --key=pki/ca.key -n kube-system
----
+
[NOTE]
====
If you want to use custom trusted CA certificate/key to sign the certificate, upload to the secret resource.

[source,bash]
----
kubectl create secret tls custom-trusted-ca --cert=<CUSTOM_TRUSTED_CA_CERT> --key=<CUSTOM_TRUSTED_CA_KEY> -n kube-system
----
====

. Install reloader by helm chart:
+
[source,bash]
----
helm install suse/reloader \
    --name ${RELEASE_NAME} \
    --namespace cert-manager
----

. Install cert-manager by helm chart:
+
[source,bash]
----
helm install jetstack/cert-manager \
    --name ${RELEASE_NAME} \
    --namespace cert-manager \
    --set global.leaderElection.namespace=cert-manager \
    --set installCRDs=true
----

> * Cert-Manager CA Issuer Resource

Reviewer: ```suggestion
* Create a Cert-Manager CA Issuer Resource
```

+

> The cert-manager CA issuer is a Kubernetes resource that represent certificate authority (CA) that are able to generate signed certificates by honoring certificate signing requests (CSR). The cert-manager certificates require a referenced issuer that is in a ready condition to attempt to honor the request.

Reviewer: ```suggestion
The cert-manager CA issuer is a Kubernetes resource that represents a certificate authority (CA) that is able to generate signed certificates by honoring certificate signing requests (CSR). Each cert-manager certificate resource requires one referenced issuer in the ready state to attempt to honor CSR requests.
```

+
[NOTE]
====

> An `Issuer` is a namespaced resource, and it is not possible to issue certificates from an Issuer in a different namespace. This means you will need to create an Issuer in each namespace you wish to obtain Certificates in.

Reviewer: ```suggestion
An `Issuer` is a namespaced resource, and it can not issue certificates to the `Certificate` resources in different namespaces.
```

> If you want to create a single Issuer that can be consumed in multiple namespaces, you should consider creating a `ClusterIssuer` resource. This is almost identical to the Issuer resource, however is non-namespaced so it can be used to issue Certificates across all namespaces.

Reviewer: ```suggestion
If you want to create a single Issuer that can be consumed in multiple namespaces, you should consider creating a `ClusterIssuer` resource. This is almost identical to the Issuer resource, however, it is cluster-wide so it can be used to issue certificates in all namespaces.
```

====
+

> Create a CA issuer called `kubernetes-ca` that will sign certificates based on a private key. The certificate stored in the secret `kubernetes-ca` can then be used to trust newly signed certificates by the kubernetes-ca issuer.

Reviewer: ```suggestion
Create a CA issuer called `kubernetes-ca` that will sign incoming certificate requests based on the CA certificate and private key stored in the secret `kubernetes-ca` used to trust newly signed certificates.
```

+
[source,bash]
----
cat << EOF > issuer.yaml
apiVersion: cert-manager.io/v1alpha3
kind: Issuer
metadata:
  name: kubernetes-ca // <1>
  namespace: kube-system
spec:
  ca:
    secretName: kubernetes-ca // <2>
EOF

kubectl apply -f issuer.yaml
----
<1> The issuer name.
<2> The secret reference name.
+
[NOTE]
====
If you want to use custom trusted CA certificate/key to sign the certificate, create a custom trust CA issuer.

[source,bash]
----
cat << EOF > custom-trusted-ca-issuer.yaml
apiVersion: cert-manager.io/v1alpha3
kind: Issuer // <1>
metadata:
  name: custom-trusted-ca
  namespace: kube-system
spec:
  ca:
    secretName: custom-trusted-ca
EOF

kubectl apply -f custom-trusted-ca-issuer.yaml
----
<1> Issuer or ClusterIssuer.
====

> * Cert-Manager Certificate Resource

Reviewer: ```suggestion
* Create a Cert-Manager Certificate Resource
```

+

> The cert-manager has the concept of Certificates that define a desired x509 certificate which will be renewed and kept up to date.

Reviewer: ```suggestion
The cert-manager has a custom resource, Certificate, which can be used to define a requested x509 certificate which will be renewed and kept up to date by an Issuer or ClusterIssuer resource.
```

kind: Certificate
metadata:
  name: oidc-dex-cert
  namespace: kube-system
spec:
  subject:
    organizations:
    - system:masters
  commonName: oidc-dex
  duration: 8760h # 1 year // <1>
  renewBefore: 720h # 1 month // <2>
  # At least one of a DNS Name or IP address is required.
  dnsNames:
  - $(cat admin.conf | grep server | awk '{print $2}' | sed 's/https:\/\///g' | sed 's/:6443//g') // <3>
  ipAddresses:
  - $(cat admin.conf | grep server | awk '{print $2}' | sed 's/https:\/\///g' | sed 's/:6443//g') // <4>
  secretName: oidc-dex-cert
  issuerRef:
    name: kubernetes-ca // <5>
    kind: Issuer // <6>
    group: cert-manager.io
  isCA: false
  usages:
    - digital signature
    - key encipherment
    - server auth
  keySize: 2048
  keyAlgorithm: rsa
  keyEncoding: pkcs1
EOF

kubectl apply -f oidc-dex-certificate.yaml
----
<1> Certificate default duration, format (XhYmZs).
<2> Certificate renew before expiration duration, format (XhYmZs).
<3> DNSNames is a list of subject alt names to be used on the Certificate.
<4> IPAddresses is a list of IP addresses to be used on the Certificate.
<5> The cert-manager issuer name.
<6> Issuer or ClusterIssuer.
+

> This certificate will tell cert-manager to attempt to use the Issuer named kubernetes-ca to obtain a certificate key pair for the domains lists in dnsNames and ipAddresses. If successful, the resulting key and certificate will be stored in a secret named oidc-dex-cert with keys of tls.key and tls.crt respectively. 

Reviewer: ```suggestion
This certificate will tell cert-manager to attempt to use the Issuer named kubernetes-ca to obtain a certificate key pair for the domain list in dnsNames and ipAddresses. If successful, the resulting key and certificate will be stored in a secret named oidc-dex-cert with keys of tls.key and tls.crt respectively. 
```

+

> The dnsNames and ipAddressed field specifies a list of Subject Alternative Names to be associated with the certificate.

Reviewer: ```suggestion
The dnsNames and ipAddressed fields specify a list of Subject Alternative Names to be associated with the certificate.
```

+

> The referenced Issuer must exist in the same namespace as the Certificate. A Certificate can alternatively reference a ClusterIssuer which is non-namespaced and so can be referenced from any namespace.

Reviewer: ```suggestion
The referenced Issuer must exist in the same namespace as the Certificate. A Certificate can alternatively reference a ClusterIssuer which is cluster-wide so it can be referenced from any namespace.
```
